### ISSUE TITLE
Enable Range type to represent all valid i64 numbers

### ISSUE THREAD

**Issue by geo-ant** [qdrant/qdrant#2356]
**Is your feature request related to a problem? Please describe.**
Hi, this relates to [this issue](https://github.com/qdrant/rust-client/issues/59) opened by another person on the rust-client. In that issue they describe that they'd like the `grpc::qdrant::Range` type to work on `i64` and `f64` in accordance with the documentation. As of now the `grpc::qdrant::Range` type only supports optional `f64` bounds. This can become a problem because there are integer values that are pretty large but well below the i64 limit that [cannot be represented](https://en.wikipedia.org/wiki/Double-precision_floating-point_format#Execution_speed_with_double-precision_arithmetic) by `f64`. I know I've actually run into this very problem when doing image processing, but whether this will ever be a problem in practice for qdrant, I don't know.

**Describe the solution you'd like**
I'd like for the `Range` type to be able to represent `i64` or `f64` ranges. As a first step one could try to restrict clients to pass only integer numbers that can actually be represented by double precision floats.

At this point I don't understand the codebase well enough to see if this is ever going to be a problem, but there seem to be at least a few places where an `i64` is cast into an `f64` (e.g. in `segment::query_optimization::condition_converter::get_range_checkers` or by implementing the `Numericable` trait on `i64`. If this is a problem I'll gladly dig in a little more and think of useful solutions.

**Describe alternatives you've considered**
Leaving as is, updating the documentation to only allow `f64` ranges.

**Additional context**
See the linked issue on the rust client.

Also please let me know if I misunderstood anything. As I said, new to the codebase so my assessment could be way off here :D

**Comment by generall** [comment]
Hm, would it solve the problem if we allow `f128`?

**Comment by geo-ant** [comment]
Definitely double check with someone else, but from what [I can see](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format) the significand precision of 113 bits should make it so that all `i64` numbers can be covered. I can only speculate on what difference it would make between using `f128` vs an enum that discriminates between `i64` and `f64`. Storage wise at least it won't matter...

**Comment by generall** [comment]
it feels like f128 would be simpler to implement. But it requires some additional checks

**Comment by geo-ant** [comment]
> it feels like f128 would be simpler to implement. But it requires some additional checks

Agree, given a `f128` type the solution would be easier (and simpler)  to implement than an enum. Just curious, what additional checks to you mean?

The one caveat is that if you ever want to extend integers to the `i128` range, you'll hit the same problem as before. Though I have no idea if `i128` integers are even a use-case that will reasonably ever pop up...

**Comment by generall** [comment]
> what additional checks to you mean?

there might be problems with migrating gRPC api

**Comment by geo-ant** [comment]
> > what additional checks to you mean?
> 
> there might be problems with migrating gRPC api

I don't believe grpc supports f128 out of the box, but I don't think Rust does either. There seem to be very few crates that are concerned with 128 bit float support. The easiest to me seems to be to roundly reject integers that cannot be accurately represented at the system boundary. Or maybe just to expose an API that accepts float64 and no i64. Alternatively, but not as simple, to refactor to a variant based solution...
